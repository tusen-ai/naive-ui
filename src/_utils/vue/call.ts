type AnyFunction = (...args: any[]) => any

function call (funcs: MaybeArray<() => void>): void
function call<A1> (funcs: MaybeArray<(a1: A1) => void>, a1: A1): void
function call<A1, A2> (
  funcs: MaybeArray<(a1: A1, a2: A2) => void>,
  a1: A1,
  a2: A2
): void
function call<A1, A2, A3> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3) => void>,
  a1: A1,
  a2: A2,
  a3: A3
): void
function call<A1, A2, A3, A4> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3, a4: A4) => void>,
  a1: A1,
  a2: A2,
  a3: A3,
  a4: A4
): void
function call<A extends any[]> (
  funcs: AnyFunction[] | AnyFunction,
  ...args: A
): void {
  if (Array.isArray(funcs)) {
    funcs.forEach((func) => (call as any)(func, ...args))
  } else return funcs(...args)
}

export { call }

export type MaybeArray<T> = T | T[]
