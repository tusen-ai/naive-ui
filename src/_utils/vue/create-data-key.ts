export function createDataKey(key: string | number): string {
  return typeof key === 'string' ? `s-${key}` : `n-${key}`
}
