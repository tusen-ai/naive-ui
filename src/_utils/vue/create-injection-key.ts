import type { InjectionKey } from 'vue'

export function createInjectionKey<T>(key: string): InjectionKey<T> {
  return key as any
}
