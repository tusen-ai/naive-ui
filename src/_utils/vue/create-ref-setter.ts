import type { Ref } from 'vue'

export function createRefSetter (ref: Ref<HTMLElement | null>): any {
  return (inst: { $el: HTMLElement | null } | null) => {
    if (inst) {
      ref.value = inst.$el
    } else {
      ref.value = null
    }
  }
}
