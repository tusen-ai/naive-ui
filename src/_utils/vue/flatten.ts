import {
  Fragment,
  type VNodeChild,
  createTextVNode,
  type VNode,
  Comment
} from 'vue'

// o(n) flatten
export function flatten (
  vNodes: VNodeChild[],
  filterCommentNode: boolean = true,
  result: VNode[] = []
): VNode[] {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    if (typeof vNode !== 'object') {
      if (typeof vNode === 'string' || typeof vNode === 'number') {
        result.push(createTextVNode(String(vNode)))
      }
      return
    }
    if (Array.isArray(vNode)) {
      flatten(vNode, filterCommentNode, result)
      return
    }
    if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children)) {
        flatten(vNode.children, filterCommentNode, result)
      }
      // rawSlot
    } else if (vNode.type !== Comment) {
      result.push(vNode)
    }
  })
  return result
}
