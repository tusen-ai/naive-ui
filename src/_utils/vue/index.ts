export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export { getFirstSlotVNode } from './get-first-slot-vnode'
export { createDataKey } from './create-data-key'
export { createRefSetter } from './create-ref-setter'
export { createInjectionKey } from './create-injection-key'
export {
  resolveSlot,
  resolveWrappedSlot,
  resolveSlotWithProps,
  resolveWrappedSlotWithProps,
  isSlotEmpty
} from './resolve-slot'
export type { MaybeArray } from './call'
export { mergeEventHandlers } from './merge-handlers'
export { isNodeVShowFalse } from './is-node-v-show-false'
export { Wrapper } from './wrapper'
