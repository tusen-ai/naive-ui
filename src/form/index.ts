export { formProps, default as NForm } from './src/Form'
export type { FormProps } from './src/Form'
export { formItemProps, default as NFormItem } from './src/FormItem'
export type { FormItemProps } from './src/FormItem'
// deprecated
export { default as NFormItemCol } from './src/FormItemCol'
export type { FormItemColProps } from './src/FormItemCol'
export {
  formItemGiProps,
  formItemGiProps as formItemGridItemProps,
  default as NFormItemGi,
  default as NFormItemGridItem
} from './src/FormItemGridItem'

export type {
  FormItemGiProps,
  FormItemGiProps as FormItemGridItemProps
} from './src/FormItemGridItem'
export { default as NFormItemRow } from './src/FormItemRow'
export type { FormItemRowProps } from './src/FormItemRow'
export type {
  FormInst,
  FormItemInst,
  FormItemRule,
  FormRules,
  FormValidationError
} from './src/interface'
