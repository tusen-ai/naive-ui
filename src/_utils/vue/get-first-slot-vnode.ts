import { type Slots, type VNode } from 'vue'
import { flatten } from './flatten'
import { warn } from '../naive'

export function getFirstSlotVNode (
  slots: Slots,
  slotName = 'default',
  props: unknown = undefined
): VNode | null {
  const slot = slots[slotName]
  if (!slot) {
    warn('getFirstSlotVNode', `slot[${slotName}] is empty`)
    return null
  }
  const slotContent = flatten(slot(props))
  // vue will normalize the slot, so slot must be an array
  if (slotContent.length === 1) {
    return slotContent[0]
  } else {
    warn('getFirstSlotVNode', `slot[${slotName}] should have exactly one child`)
    return null
  }
}
