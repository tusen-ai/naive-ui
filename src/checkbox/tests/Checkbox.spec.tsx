import { h } from 'vue'
import { mount, VueWrapper } from '@vue/test-utils'
import { NCheckbox, NCheckboxGroup } from '../index'
import { NForm, NFormItem } from '../../form'

function expectChecked (wrapper: VueWrapper<any>, value: boolean): void {
  expect(wrapper.classes().some((c) => c.includes('checked'))).toEqual(value)
  expect(wrapper.find('.n-checkbox').attributes('aria-checked')).toBe(
    value.toString()
  )
}

describe('n-checkbox', () => {
  it('should work with import on demand', () => {
    mount(NCheckbox)
  })

  describe('uncontrolled mode', () => {
    it('works', async () => {
      const wrapper = mount(NCheckbox)
      expectChecked(wrapper, false)
      await wrapper.trigger('click')
      expectChecked(wrapper, true)
      await wrapper.trigger('click')
      expectChecked(wrapper, false)
    })
    it('props.defaultChecked', () => {
      const wrapper = mount(NCheckbox, {
        props: {
          defaultChecked: true
        }
      })
      expectChecked(wrapper, true)
    })
  })

  it('should work with `checked-value` prop', async () => {
    const onUpdateChecked = vi.fn()
    const wrapper = mount(NCheckbox, {
      props: {
        checkedValue: 'fooo',
        uncheckedValue: 'barr',
        onUpdateChecked
      }
    })
    await wrapper.trigger('click')
    expect(onUpdateChecked.mock.calls[0][0]).toEqual('fooo')
    await wrapper.trigger('click')
    expect(onUpdateChecked.mock.calls[1][0]).toEqual('barr')
    await wrapper.trigger('click')
    expect(onUpdateChecked.mock.calls[2][0]).toEqual('fooo')
  })

  it('should work with `checked-value` prop in type layer', () => {
    const onUpdateChecked1: (value: string) => void = () => {}
    const onUpdateChecked2: (value: number) => void = () => {}
    const onUpdateChecked3: (value: boolean) => void = () => {}
    let _ = (
      <NCheckbox
        onUpdateChecked={onUpdateChecked1}
        checked={'123'}
        defaultChecked={'123'}
      />
    )
    _ = (
      <NCheckbox
        onUpdateChecked={onUpdateChecked2}
        checked={123}
        defaultChecked={123}
      />
    )
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _ = (
      <NCheckbox
        onUpdateChecked={onUpdateChecked3}
        checked={true}
        defaultChecked={false}
      />
    )
  })

  it('should work with `indeterminate` prop', () => {
    const wrapper = mount(NCheckbox, {
      props: {
        indeterminate: true
      },
      slots: {
        default: () => 'test'
      }
    })
    expect(wrapper.find('.n-checkbox').classes()).toContain(
      'n-checkbox--indeterminate'
    )
    expect(wrapper.find('.n-checkbox').attributes('aria-checked')).toBe('mixed')
  })

  it('should work with `disabled` prop', () => {
    const wrapper = mount(NCheckbox, {
      props: {
        disabled: true
      },
      slots: {
        default: () => 'test'
      }
    })
    expect(wrapper.find('.n-checkbox').classes()).toContain(
      'n-checkbox--disabled'
    )
  })

  it('should work with `focusable` prop', async () => {
    const wrapper = mount(NCheckbox, {
      props: {
        focusable: false
      },
      slots: {
        default: () => 'test'
      }
    })
    expect(wrapper.find('[tabindex]').exists()).not.toBe(true)

    await wrapper.setProps({ focusable: true })
    expect(wrapper.find('[tabindex]').exists()).toBe(true)
    expect(wrapper.find('.n-checkbox').attributes('tabindex')).toContain('0')
  })

  it('should work with `label` prop', async () => {
    const wrapper = mount(NCheckbox, {
      props: {
        label: 'test'
      }
    })
    expect(wrapper.find('.n-checkbox__label').text()).toContain('test')
  })

  it('should work with `on-update:checked` & `onUpdateChecked` prop', async () => {
    const onClick = vi.fn()
    const wrapper = mount(NCheckbox, {
      props: {
        'onUpdate:checked': onClick,
        onUpdateChecked: onClick
      },
      slots: {
        default: () => 'test'
      }
    })

    await wrapper.trigger('click')
    expect(onClick).toHaveBeenCalledTimes(2)
  })

  it('should work with default slots', async () => {
    const wrapper = mount(NCheckbox, {
      slots: {
        default: () => 'test'
      }
    })

    expect(wrapper.find('.n-checkbox__label').text()).toContain('test')
  })

  it('should work with `size` prop', () => {
    ;(['small', 'medium', 'large'] as const).forEach((i) => {
      const wrapper = mount(NCheckbox, { props: { size: i } })
      expect(wrapper.find('.n-checkbox').attributes('style')).toMatchSnapshot()
    })
  })

  it('should show correct style with `NForm` component', () => {
    const wrapper = mount(() => (
      <NForm size="medium">
        {{
          default: () => {
            return (
              <NFormItem>
                {{
                  default: () => <NCheckbox />
                }}
              </NFormItem>
            )
          }
        }}
      </NForm>
    ))

    expect(wrapper.find('.n-checkbox').attributes('style')).toMatchSnapshot()
  })

  describe('accessibility', () => {
    it('should have a role of "checkbox"', () => {
      const wrapper = mount(NCheckbox)
      expect(wrapper.find('.n-checkbox').attributes('role')).toBe('checkbox')
    })

    it('should set a default aria-labelledby', () => {
      const labelId = 'custom-id'
      const wrapper = mount(() => <NCheckbox aria-labelledby={labelId} />)
      expect(wrapper.find('.n-checkbox').attributes('aria-labelledby')).toMatch(
        labelId
      )
    })

    it('should allow to set aria-labelledby from outside', () => {
      const wrapper = mount(NCheckbox)
      const labelId = wrapper.find('.n-checkbox__label').attributes('id')
      expect(wrapper.find('.n-checkbox').attributes('aria-labelledby')).toBe(
        labelId
      )
    })
  })
})

describe('n-checkbox-group', () => {
  it('should work with import on demand', () => {
    mount(NCheckboxGroup)
  })

  it('should have a role of "group"', () => {
    const wrapper = mount(NCheckboxGroup)
    expect(wrapper.find('.n-checkbox-group').attributes('role')).toBe('group')
  })

  it('should work with `disabled` prop', () => {
    const wrapper = mount(NCheckboxGroup, {
      props: {
        disabled: true
      },
      slots: {
        default: () => h(NCheckbox, null, { default: () => 'test' })
      }
    })
    expect(wrapper.find('.n-checkbox--disabled').exists()).toBe(true)
  })

  it('should work with `on-update:value` prop', async () => {
    const onClick = vi.fn()
    const wrapper = mount(NCheckboxGroup, {
      props: {
        'on-update:value': onClick
      },
      slots: {
        default: () => h(NCheckbox, { value: 'test' })
      }
    })
    await wrapper.findComponent(NCheckbox).trigger('click')
    expect(onClick).toBeCalled()
  })

  it('should work with default slots', async () => {
    const wrapper = mount(NCheckboxGroup, {
      props: {
        disabled: true
      },
      slots: {
        default: () => h(NCheckbox, null, { default: () => 'test' })
      }
    })
    expect(wrapper.find('.n-checkbox__label').text()).toContain('test')
  })

  it('should work with `min` prop', async () => {
    const wrapper = mount(NCheckboxGroup, {
      props: {
        min: 1,
        value: ['Shanghai']
      },
      slots: {
        default: () => [
          h(NCheckbox, { value: 'Shanghai' }, { default: () => 'Shanghai' }),
          h(NCheckbox, { value: 'Beijing' }, { default: () => 'Beijing' }),
          h(NCheckbox, { value: 'Shenzhen' }, { default: () => 'Shenzhen' })
        ]
      }
    })
    expect(wrapper.findAll('.n-checkbox').length).toBe(3)

    expect(wrapper.findAll('.n-checkbox')[0].classes()).toContain(
      'n-checkbox--disabled'
    )
    expect(wrapper.findAll('.n-checkbox')[0].classes()).toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[1].classes()).not.toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[1].classes()).not.toContain(
      'n-checkbox--disabled'
    )
  })

  it('should work with `max` prop', async () => {
    const wrapper = mount(NCheckboxGroup, {
      props: {
        max: 2,
        value: ['Shanghai', 'Beijing']
      },
      slots: {
        default: () => [
          h(NCheckbox, { value: 'Shanghai' }, { default: () => 'Shanghai' }),
          h(NCheckbox, { value: 'Beijing' }, { default: () => 'Beijing' }),
          h(NCheckbox, { value: 'Shenzhen' }, { default: () => 'Shenzhen' })
        ]
      }
    })
    expect(wrapper.findAll('.n-checkbox').length).toBe(3)

    expect(wrapper.findAll('.n-checkbox')[0].classes()).not.toContain(
      'n-checkbox--disabled'
    )
    expect(wrapper.findAll('.n-checkbox')[0].classes()).toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[1].classes()).toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[1].classes()).not.toContain(
      'n-checkbox--disabled'
    )
    expect(wrapper.findAll('.n-checkbox')[2].classes()).not.toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[2].classes()).toContain(
      'n-checkbox--disabled'
    )
  })

  it('should work with `max` and `min` prop', async () => {
    const wrapper = mount(NCheckboxGroup, {
      props: {
        max: 2,
        min: 1
      },
      slots: {
        default: () => [
          h(NCheckbox, { value: 'Shanghai' }, { default: () => 'Shanghai' }),
          h(NCheckbox, { value: 'Beijing' }, { default: () => 'Beijing' }),
          h(NCheckbox, { value: 'Shenzhen' }, { default: () => 'Shenzhen' })
        ]
      }
    })

    await wrapper.setProps({
      value: ['Shanghai']
    })
    expect(wrapper.findAll('.n-checkbox').length).toBe(3)

    expect(wrapper.findAll('.n-checkbox')[0].classes()).toContain(
      'n-checkbox--disabled'
    )
    expect(wrapper.findAll('.n-checkbox')[0].classes()).toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[1].classes()).not.toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[1].classes()).not.toContain(
      'n-checkbox--disabled'
    )
    expect(wrapper.findAll('.n-checkbox')[2].classes()).not.toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[2].classes()).not.toContain(
      'n-checkbox--disabled'
    )
    await wrapper.setProps({
      value: ['Shanghai', 'Beijing']
    })

    expect(wrapper.findAll('.n-checkbox')[0].classes()).not.toContain(
      'n-checkbox--disabled'
    )
    expect(wrapper.findAll('.n-checkbox')[0].classes()).toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[1].classes()).toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[1].classes()).not.toContain(
      'n-checkbox--disabled'
    )
    expect(wrapper.findAll('.n-checkbox')[2].classes()).not.toContain(
      'n-checkbox--checked'
    )
    expect(wrapper.findAll('.n-checkbox')[2].classes()).toContain(
      'n-checkbox--disabled'
    )
  })
})
