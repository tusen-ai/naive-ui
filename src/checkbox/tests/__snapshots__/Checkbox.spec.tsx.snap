// Vitest Snapshot v1

exports[`n-checkbox > should show correct style with \`NForm\` component 1`] = `"--n-label-line-height: 1.6; --n-size: 16px; --n-bezier: cubic-bezier(.4, 0, .2, 1); --n-border-radius: 2px; --n-border: 1px solid rgb(224, 224, 230); --n-border-checked: 1px solid #18a058; --n-border-focus: 1px solid #18a058; --n-border-disabled: 1px solid rgb(224, 224, 230); --n-border-disabled-checked: 1px solid rgb(224, 224, 230); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.3); --n-color: #FFF; --n-color-checked: #18a058; --n-color-table: #fff; --n-color-table-modal: #fff; --n-color-table-popover: #fff; --n-color-disabled: rgb(250, 250, 252); --n-color-disabled-checked: rgb(250, 250, 252); --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-check-mark-color: #FFF; --n-check-mark-color-disabled: rgba(194, 194, 194, 1); --n-check-mark-color-disabled-checked: rgba(194, 194, 194, 1); --n-font-size: 14px; --n-label-padding: 0 8px;"`;

exports[`n-checkbox > should work with \`size\` prop 1`] = `"--n-label-line-height: 1.6; --n-size: 14px; --n-bezier: cubic-bezier(.4, 0, .2, 1); --n-border-radius: 2px; --n-border: 1px solid rgb(224, 224, 230); --n-border-checked: 1px solid #18a058; --n-border-focus: 1px solid #18a058; --n-border-disabled: 1px solid rgb(224, 224, 230); --n-border-disabled-checked: 1px solid rgb(224, 224, 230); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.3); --n-color: #FFF; --n-color-checked: #18a058; --n-color-table: #fff; --n-color-table-modal: #fff; --n-color-table-popover: #fff; --n-color-disabled: rgb(250, 250, 252); --n-color-disabled-checked: rgb(250, 250, 252); --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-check-mark-color: #FFF; --n-check-mark-color-disabled: rgba(194, 194, 194, 1); --n-check-mark-color-disabled-checked: rgba(194, 194, 194, 1); --n-font-size: 14px; --n-label-padding: 0 8px;"`;

exports[`n-checkbox > should work with \`size\` prop 2`] = `"--n-label-line-height: 1.6; --n-size: 16px; --n-bezier: cubic-bezier(.4, 0, .2, 1); --n-border-radius: 2px; --n-border: 1px solid rgb(224, 224, 230); --n-border-checked: 1px solid #18a058; --n-border-focus: 1px solid #18a058; --n-border-disabled: 1px solid rgb(224, 224, 230); --n-border-disabled-checked: 1px solid rgb(224, 224, 230); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.3); --n-color: #FFF; --n-color-checked: #18a058; --n-color-table: #fff; --n-color-table-modal: #fff; --n-color-table-popover: #fff; --n-color-disabled: rgb(250, 250, 252); --n-color-disabled-checked: rgb(250, 250, 252); --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-check-mark-color: #FFF; --n-check-mark-color-disabled: rgba(194, 194, 194, 1); --n-check-mark-color-disabled-checked: rgba(194, 194, 194, 1); --n-font-size: 14px; --n-label-padding: 0 8px;"`;

exports[`n-checkbox > should work with \`size\` prop 3`] = `"--n-label-line-height: 1.6; --n-size: 18px; --n-bezier: cubic-bezier(.4, 0, .2, 1); --n-border-radius: 2px; --n-border: 1px solid rgb(224, 224, 230); --n-border-checked: 1px solid #18a058; --n-border-focus: 1px solid #18a058; --n-border-disabled: 1px solid rgb(224, 224, 230); --n-border-disabled-checked: 1px solid rgb(224, 224, 230); --n-box-shadow-focus: 0 0 0 2px rgba(24, 160, 88, 0.3); --n-color: #FFF; --n-color-checked: #18a058; --n-color-table: #fff; --n-color-table-modal: #fff; --n-color-table-popover: #fff; --n-color-disabled: rgb(250, 250, 252); --n-color-disabled-checked: rgb(250, 250, 252); --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-check-mark-color: #FFF; --n-check-mark-color-disabled: rgba(194, 194, 194, 1); --n-check-mark-color-disabled-checked: rgba(194, 194, 194, 1); --n-font-size: 15px; --n-label-padding: 0 8px;"`;
